packages/react-admin/src/AdminRoutes.js

```diff
@@ -7,22 +7,44 @@ import getContext from 'recompose/getContext';
 
 import CrudRoute from './CrudRoute';
 import NotFound from './mui/layout/NotFound';
-import Restricted from './auth/Restricted';
-import { AUTH_GET_PERMISSIONS } from './auth';
+import WithPermissions from './auth/WithPermissions';
 import { declareResources as declareResourcesAction } from './actions';
+import { AUTH_GET_PERMISSIONS } from './auth/types';
+
+const initialPermissions = '@@ar/initialPermissions';
 
 export class AdminRoutes extends Component {
+    // Can't use null or undefined here as authClient may return any those values
+    state = { permissions: initialPermissions };
+
     componentDidMount() {
-        return this.initializeResources(this.props.children);
+        this.getPermissions();
     }
 
-    async initializeResources(children) {
-        if (typeof children === 'function') {
-            let permissions;
-            if (this.props.authClient) {
-                permissions = await this.props.authClient(AUTH_GET_PERMISSIONS);
-            }
+    componentDidUpdate(prevProps, prevState) {
+        if (
+            prevState.permissions !== this.state.permissions &&
+            this.state.permissions !== initialPermissions
+        ) {
+            this.initializeResources();
+        }
+    }
 
+    getPermissions = async () => {
```

Comment:
As this is the job of `WithPermissions`, why don't we decorate `<AdminRoutes>` with `<WithPermissions>` instead of duplicating the code here?

Reply:
Because `AdminRoutes` is also responsible for rendering custom routes which may not be protected and `WithPermissions` will check for authentication